[BLOG] Added blog post for enhancing developer experience at SciPy - Intel oneAPI, MSVC and spin support

czgdp1807 opened the pull request:
<!--Hi there! If you aren't submitting a blog post, you can delete this message.
Thank you for submitting a blog post! We want to make sure content produced for Quansight Labs is accessible, so please address the the following guidelines in your post.-->

## Text styling

- [x] The blog is written with [plain language](https://www.plainlanguage.gov/guidelines/) (where relevant).
- [x] If there are headers, they use the proper header tags in order to do so (with only one level-one header).
- [x] All links describe where they link to (for example, check the [Quansight labs website](https://labs.quansight.org/)).
- [x] Any kind of styling that the author uses (for example, bold for emphasis) is consistent throughout the blog.

## Non-text contents

- [x] Blog post featured image is in PNG or JPEG format, **not** SVG.
- [x] All content is represented as text (for example, images need alt text and videos need captions or descriptive transcripts).
- [x] If there are emojis, there are not more than three in a row.
- [x] Don't use [flashing gifs or videos](https://www.w3.org/TR/UNDERSTANDING-WCAG20/seizure-does-not-violate.html).
- [x] If it were to be read as plain text, the blog still makes sense and no information is missing.

czgdp1807 commented:
The blog preview is https://labs-ksnzv1tzf-quansight.vercel.app/blog/intel-oneapi-msvc-spin-scipy. All the content is as it is (`asciinema` videos are also there in their respective places).

czgdp1807 commented on apps/labs/posts/intel-oneapi-msvc-spin-scipy.md:
```diff
@@ -0,0 +1,88 @@
+---
+title: 'Enhancing Developer Experience at SciPy - Intel oneAPI/MSVC Support and Migrating to `spin`'
```
Should I remove, "\`" from "`spin`"?

czgdp1807 reviewed (commented):
Also, I have SciPy logo as featured and hero images. I hope that should work because its 100% SciPy work.

czgdp1807 commented:
I have shortened the links. However at some places (very few) I had to use PR titles so I put the title inside "\`" and linked the issue \[\[pr_number\]\]\(pr_link\). That way the link is short. Please let me know if any other changes are needed.

czgdp1807 commented:
I checked the preview - https://labs-pu7mgao24-quansight.vercel.app - everything loads correctly (`asciinema` videos and the text of the blog). See the screenshots below. I have zoomed out to cover the whole page in screenshot.

<details>

<summary>Screenshots of `asciinema` videos loading correctly</summary>

<img width="1352" alt="Screenshot 2024-12-22 at 12 35 33 PM" src="https://github.com/user-attachments/assets/01b3e851-1a00-43dc-ae7c-4151cc85b7ba" />
<img width="1352" alt="Screenshot 2024-12-22 at 12 35 45 PM" src="https://github.com/user-attachments/assets/8dd60f16-a7f5-4cb8-a31b-067d67723e28" />
<img width="1352" alt="Screenshot 2024-12-22 at 12 35 56 PM" src="https://github.com/user-attachments/assets/4f8a3c50-66de-4012-a51c-7489201b2c91" />
<img width="1352" alt="Screenshot 2024-12-22 at 12 36 10 PM" src="https://github.com/user-attachments/assets/2a7f4faf-8c4a-4e5a-87c7-7b831180f8a7" />
<img width="1352" alt="Screenshot 2024-12-22 at 12 36 22 PM" src="https://github.com/user-attachments/assets/536241a6-121a-41ca-ab70-a30c4e63dbd3" />

</details>

rgommers commented on apps/labs/posts/intel-oneapi-msvc-spin-scipy.md:
```diff
@@ -0,0 +1,88 @@
+
+<script src="https://asciinema.org/a/aZstdaf6B6nlkW9JFccT7z7Dr.js" id="asciicast-aZstdaf6B6nlkW9JFccT7z7Dr" async="true"></script>
+
+The above video showcases how `spin` works fine with Homebrew installed Python 3.10. The build completes. However, `dev.py` fails because it is unable to resolve paths correctly.
```
It would be good to add acknowledgements - this was done with NASA funding. Can you please copy the phrasing at the end of Albert's recent blog post for this?

rgommers commented on apps/labs/posts/intel-oneapi-msvc-spin-scipy.md:
```diff
@@ -0,0 +1,88 @@
+
+#### Intel oneAPI Support in SciPy
+
+Intel offers oneAPI - a complete toolkit for HPC use cases. It has ICX (for C), ICPX (for C++), IFX (for Fortran) and MKL (an OpenBLAS alternative). We targeted two operating systems, Windows and Linux. `CI: adding a Windows CI job with MSVC + MKL + Intel Fortran (ifx)` [[gh-20878](https://github.com/scipy/scipy/issues/20878)] was the starting point. Solving this issue meant, avoiding regressions ...
```
This "It has ICX ..." is repeated from a few paragraphs up. I actually suggest moving that content here.

rgommers commented on apps/labs/posts/intel-oneapi-msvc-spin-scipy.md:
```diff
@@ -0,0 +1,88 @@
+1. Intel oneAPI and MSVC Support in SciPy
+2. Moving away from `dev.py` and migrating to `spin`.
+
+First, let me share the motivation behind these.
```
It would be better I think to share the general motivation in fewer words. In-depth paragraphs about both topics before repeating part of the in-depth content in the separate sections on OneAPI and `spin` feels a little off.

rgommers reviewed (commented):
Thanks @czgdp1807. A set of inline comments. Can you please line-break the content at 80 chars or so? That's way easier to review line by line.

rgommers commented:
Sorry, only half a review, hit submit too early - more comments coming.

rgommers commented on apps/labs/posts/intel-oneapi-msvc-spin-scipy.md:
```diff
@@ -0,0 +1,88 @@
+
+First, let me share the motivation behind these.
+
+Intel offers a complete toolkit for high performance computing (HPC), numerical computing use cases. It goes by the name of [oneAPI](https://www.intel.com/content/www/us/en/developer/tools/oneapi/overview.html). It provides, ICX (for C), ICPX (for C++), IFX (for Fortran) and Math Kernal Library (MKL for optimized math routines like BLAS, LAPACK, fast FFT, etc.) [[1]](https://en.wikipedia.org/wiki ...
```
- I'd use "MKL (Math Kernel Library, ...)" rather than the other way around, since you're also using the abbreviations for ICX & co
- MSVC: we didn't actually add support for it though, so the rationale here is a little off. Try to rephrase to what we actually aimed to do.

czgdp1807 replied:
I think MSVC wasn't tested on CI before our work, right? On Windows we used clang for CI testing. So may be a better line can be, "Since it is the de-facto compiler for C/C++ on Windows, we added CI jobs for testing SciPy compilation with MSVC"

rgommers commented on apps/labs/posts/intel-oneapi-msvc-spin-scipy.md:
```diff
@@ -0,0 +1,88 @@
+
+Intel offers a complete toolkit for high performance computing (HPC), numerical computing use cases. It goes by the name of [oneAPI](https://www.intel.com/content/www/us/en/developer/tools/oneapi/overview.html). It provides, ICX (for C), ICPX (for C++), IFX (for Fortran) and Math Kernal Library (MKL for optimized math routines like BLAS, LAPACK, fast FFT, etc.) [[1]](https://en.wikipedia.org/wiki ...
+
+Now let's see why we decided to spin from `dev.py` to `spin` (pun inteded). `dev.py` (or `do.py` [[2]](https://labs.quansight.org/blog/the-evolution-of-the-scipy-developer-cli)) was added to improve the experience for SciPy developers. Basically, it offers better documentation for the various commands needed (like, `build`, `test`, `shell`) while working on SciPy codebase. It has better looking C ...
```
I'd say the long-term maintenance load sharing and the familiarity for contributors coming from `numpy` or other project are the main motivations. The paper cuts & bugs are very minor.

rgommers replied:
Also, it'd be good to say somehting like "`dev.py` was a major improvement in development workflow, and came before `spin`. `spin` took its inspiration [in large part](https://github.com/scientific-python/spin/?tab=readme-ov-file#history) from `dev.py`, and generalized its interface to work with more projects without customization as heavy as `dev.py` has for SciPy".

rgommers commented on apps/labs/posts/intel-oneapi-msvc-spin-scipy.md:
```diff
@@ -0,0 +1,88 @@
+
+One major issue with MSVC is that its support for handling arrays on stack (with runtime sizes) is limited. Expressions like, `std::complex<double> cwrk[n];` fail to compile with MSVC unless `n` is a constant. So we did runtime allocation on heap (using `new`) and performed `delete` at the end to free memory. This is a manual approach of handling memory, however this is the only solution which wo ...
+
+Now coming on to the `arpack` issue. Fixing this issue required supporting building SciPy with Intel oneAPI. On Linux the `arpack` issue was already resolved. However, there were some other issues with `ifx`. For example, `test_equal_bounds` test in `scipy/optimize/tests/test_optimize.py` failed due to floating point issues with `ifx`. Also tolerances for other tests had to be increased (i.e., sl ...
```
You may want to link to https://github.com/opencollab/arpack-ng, most readers won't know what ARPACK is.

rgommers commented on apps/labs/posts/intel-oneapi-msvc-spin-scipy.md:
```diff
@@ -0,0 +1,88 @@
+
+Now coming on to the `arpack` issue. Fixing this issue required supporting building SciPy with Intel oneAPI. On Linux the `arpack` issue was already resolved. However, there were some other issues with `ifx`. For example, `test_equal_bounds` test in `scipy/optimize/tests/test_optimize.py` failed due to floating point issues with `ifx`. Also tolerances for other tests had to be increased (i.e., sl ...
+
+The final checkpoint for Intel oneAPI was `CI: Test icx + icpx + ifx + MKL build of SciPy` [[gh-21254]](https://github.com/scipy/scipy/pull/21254). Here we replaced `gcc` with `icx` and `g++` with `icpx` and TADA, it worked on Linux. Regarding Windows, we are still at `MSVC` + `ifx` + `OpenBLAS` combination because of the `arpack` import error. We tried several ways to fix it but all resulted in  ...
```
The last two sentences are not a very good ending. Try to say why we stopped (e.g., "the effort wasn't justified by ..."), otherwise it leaves the reader wondering.

rgommers commented on apps/labs/posts/intel-oneapi-msvc-spin-scipy.md:
```diff
@@ -0,0 +1,88 @@
+
+Now it would be easy to understand our work below.
+
+We decided to shift to `spin` because of the several issues reported in `dev.py`. For example, [doit interfering with command history of pdb] [[gh-16452]](https://github.com/scipy/scipy/issues/16452), usage of emojis giving issues in a Windows CI environment [[gh-18046]](https://github.com/scipy/scipy/issues/18046), and dev.py unable to resolve homebrew installed Python [[gh-18998]](https://githu ...
```
repeating grammar issue, so commenting: it's "fewer bugs" rather than "lesser bugs" (check for "lesser" in the whole post please)

rgommers reviewed (commented):
It'd be useful to spell check the content - there are a few too many mistakes.

The asciinema videos are still not rendering, I don't understand what is happening in the previews. The videos also need alt text (see checklist in PR description).

czgdp1807 commented:
@rgommers Taking into account your reviews, I have done the following,

1. Improvised the language of the blog post content. Should be more lucid to read now.
2. Added short text descriptions for asciinema videos.
3. The videos are now linked to my `asciinema` account, so they should load. If not then we have the text descriptions.
4. Fixed spelling mistakes via Google docs.

Please go through it now and let me know your experience. Drop in any reviews where you think changes are needed.

Thank you.

pavithraes commented on apps/labs/posts/intel-oneapi-msvc-spin-scipy.md:
```diff
@@ -0,0 +1,174 @@
+  imageAlt: 'SciPy logo'
+---
+
+In this blog post, we’ll discuss the improvements we made to enhance the developer experience at SciPy. Our work focused on two key areas:
```
```suggestion
Over the past year, we worked on enhancing the development experience of SciPy. We focused on two key areas:
```

The start feels a little abrupt, hence this suggestion for a softer introduction.

pavithraes commented on apps/labs/posts/intel-oneapi-msvc-spin-scipy.md:
```diff
@@ -0,0 +1,174 @@
+First, let’s explore the motivation behind these updates.
+
+Intel provides a comprehensive toolkit for high-performance computing (HPC) and numerical computing use cases, known as [oneAPI](https://www.intel.com/content/www/us/en/developer/tools/oneapi/overview.html).
+Given that SciPy is heavily used in HPC and numerical computing, supporting the ability to build SciPy with Intel oneAPI was a natural and important step.
+By doing so, we aim to improve the experience for SciPy users who rely on Intel oneAPI as their primary toolkit.
```
```suggestion
SciPy is heavily used in HPC and numerical computing.
Adding support for SciPy builds with Intel oneAPI was a natural and important step to improve the experience for SciPy users who rely on Intel oneAPI as their primary toolkit.
```

pavithraes commented on apps/labs/posts/intel-oneapi-msvc-spin-scipy.md:
```diff
@@ -0,0 +1,174 @@
+Additionally, since MSVC is the de facto compiler for C/C++ on Windows, we added continuous integration (CI) jobs to test SciPy compilation with MSVC, further expanding the compatibility of SciPy across platforms.
+
+Now, let’s talk about the transition from `dev.py` to `spin`.
+`dev.py` (also known as `do.py` [[1]](https://labs.quansight.org/blog/the-evolution-of-the-scipy-developer-cli)) was a significant step forward in improving the development workflow.
```
```suggestion
`dev.py` (also known as `do.py`, learn more in [The evolution of the SciPy developer CLI](https://labs.quansight.org/blog/the-evolution-of-the-scipy-developer-cli)) was a significant step forward in improving the development workflow.
```

We use this paradigm for footnotes. For links, having descriptive text is more accessible.

pavithraes commented on apps/labs/posts/intel-oneapi-msvc-spin-scipy.md:
```diff
@@ -0,0 +1,174 @@
+#### Intel oneAPI Support in SciPy
+
+Intel offers oneAPI, a comprehensive toolkit designed for high-performance computing (HPC) applications.
+It includes essential components such as `icx` (for C), `icpx` (for C++), `ifx` (for Fortran), and MKL (Math Kernel Library), which provides optimized mathematical routines like BLAS, LAPACK, and fast FFTs [[2]](https://en.wikipedia.org/wiki/Math_Kernel_Library).
```
```suggestion
It includes essential components such as `icx` (for C), `icpx` (for C++), `ifx` (for Fortran), and [MKL (Math Kernel Library)](https://en.wikipedia.org/wiki/Math_Kernel_Library), which provides optimized mathematical routines like BLAS, LAPACK, and fast FFTs.
```

pavithraes commented on apps/labs/posts/intel-oneapi-msvc-spin-scipy.md:
```diff
@@ -0,0 +1,174 @@
+Once we addressed this issue, we added a CI job to test SciPy's compilation with MSVC + `ifx` + OpenBLAS, which helped prevent future regressions in the MSVC build.
+This also marked an early step toward supporting Intel oneAPI within SciPy.
+
+The next major hurdle involved the [`arpack`](https://github.com/opencollab/arpack-ng) issue, which required full Intel oneAPI support to resolve.
```
I don't think I understand what the "arpack issue" is? Could we link to any issue on the scipy tracker that has more context?

czgdp1807 replied:
I am referring to https://github.com/scipy/scipy/issues/20728. You are right, just linking it here as well should make it clear. I have referred to this issue in lines 42-44.

czgdp1807 replied:
Done. Let me know if it looks good to you.

pavithraes commented on apps/labs/posts/intel-oneapi-msvc-spin-scipy.md:
```diff
@@ -0,0 +1,174 @@
+This was all handled in the PR, `BUG/CI: Compile and run tests with ifx + MKL on Linux` [[gh-21173]](https://github.com/scipy/scipy/pull/21173), where I also added a CI job to test the combination of `gcc` + `g++` + `ifx` + MKL on Linux, marking another important step toward full Intel oneAPI support for SciPy.
+
+The final milestone for Intel oneAPI support was the CI job for testing SciPy with `icx`, `icpx`, `ifx`, and MKL [[gh-21254]](https://github.com/scipy/scipy/pull/21254).
+In this step, we replaced `gcc` with `icx` and `g++` with `icpx`, and the build successfully passed on Linux. However, on Windows, we were unable to proceed further due to an unresolved import error with `arpack`. Despite several attempts to fix the issue, all solutions led to dead ends. As a result, we decided to pause the effort on Windows for now, as the work required did not seem to justify t ...
```
```suggestion
In this step, we replaced `gcc` with `icx` and `g++` with `icpx`, and the build successfully passed on Linux.

However, on Windows, we were unable to proceed further due to an unresolved import error with `arpack`. Despite several attempts to fix the issue, all solutions led to dead ends. As it was turning into an unbounded amount of work, we decided to pause the effort on Windows and focus on on more scoped tasks.
```

pavithraes commented on apps/labs/posts/intel-oneapi-msvc-spin-scipy.md:
```diff
@@ -0,0 +1,174 @@
+One of the advantages of `spin` is that it has fewer dependencies, which in turn means fewer points of failure.
+While this trade-off sacrifices some of the output’s polish, it greatly improves the reliability of the development process for SciPy.
+
+**Why not just fix `dev.py`?**
+
+You might wonder, "Why not just update `dev.py` to fix these issues? Why switch to a new third-party tool like `spin`?"
```
```suggestion
You might wonder, "Why not just update `dev.py` to fix these issues? Why switch to a new third-party tool like `spin`?"
```

Removing since it feels redundant.

pavithraes commented on apps/labs/posts/intel-oneapi-msvc-spin-scipy.md:
```diff
@@ -0,0 +1,174 @@
+
+One of the key reasons is that **`spin` is already widely used by other scientific computing projects** such as `numpy`, `scikit-image`, `solcore5`, `skmisc`, `PyFVS`, `sktree`, `dipy`, and many more (refer to [scientific-python/spin#62](https://github.com/scientific-python/spin/issues/62)).
+This means improvements made to `spin` will automatically benefit SciPy.
+Additionally, any limitations we encounter in `spin` will be addressed through Pull Requests, making it a two-way connection: SciPy benefits from other projects (via `spin`), and those projects benefit from improvements made by SciPy.
```
```suggestion
Additionally, any limitations we encounter in `spin` will be addressed through upstream contributions to the project, making it a two-way connection: SciPy benefits from other projects contributing to `spin`, and those projects benefit from improvements made by the SciPy team to `spin`.
```

pavithraes commented on apps/labs/posts/intel-oneapi-msvc-spin-scipy.md:
```diff
@@ -0,0 +1,174 @@
+We’re waiting for the PR [scientific-python/spin#257](https://github.com/scientific-python/spin/pull/257) to be merged, after which we’ll be ready to fully integrate `spin` into SciPy.
+Our work has been fully tested in the SciPy CI system, ensuring its stability.
+
+Here’s a quick comparison of the SciPy build experience with `dev.py` versus `spin`:
```
The videos don't seem to rendering in the preview: https://labs-git-fork-czgdp1807-feature-intel-oneapi-spin-quansight.vercel.app/blog/intel-oneapi-msvc-spin-scipy

czgdp1807 replied:
I see, I can try https://docs.asciinema.org/manual/server/embedding/#preview-image-link for sharing the videos. `asciinema` will provide an image. If someone clicks it then the videos plays on asciinema website. That should solve the problems showing up due to embedding the player using `script` tag. Let me know if it would work.

rgommers replied:
That seems like a pragmatic solution - try it I'd say.

pavithraes replied:
I pushed up a commit to do this (please let me know if you're not comfortable with this, and I'll keep it mind for the future), and unblock merging this PR. :)

czgdp1807 replied:
No worries. Thanks for doing it. :D.

pavithraes commented on apps/labs/posts/intel-oneapi-msvc-spin-scipy.md:
```diff
@@ -0,0 +1,174 @@
+This video shows building SciPy with `spin`.
+
+As you can see there is no difference between the look and feel of `python dev.py build` and `spin build`.
+It works exactly the same. However, the issues which were there with `dev.py` are not observed with `spin`.
```
```suggestion
It works exactly the same. Moreover, the issues which were there with `dev.py` are not observed with `spin`.
```

pavithraes commented on apps/labs/posts/intel-oneapi-msvc-spin-scipy.md:
```diff
@@ -0,0 +1,174 @@
+
+This video shows `spin` keeps `pdb` command history intact.
+
+As you can see `spin` usage doesn't interfere with PDB's history. This issue was there with `dev.py`.
```
```suggestion
As you can see `spin` usage doesn't interfere with PDB's history, which was an issue with `dev.py`.
```

pavithraes commented on apps/labs/posts/intel-oneapi-msvc-spin-scipy.md:
```diff
@@ -0,0 +1,174 @@
+---
+title: 'Enhancing Developer Experience at SciPy - Intel oneAPI/MSVC Support and Migrating to spin'
+published: February 21, 2025
```
Just a note to ensure the date is updated before merging this PR.

pavithraes reviewed (changes requested):
@czgdp1807 Thanks for authoring this post, and for your patience with the review. It's a good read, and I learnt some new things! I've shared a few narrative suggestions. 

The videos not rendering in the preview is a blocker. I'm not familiar with asciinema, and need to look into it.

pavithraes reviewed (approved):
Thanks again! All the comments are addressed. :)

czgdp1807 commented:
Less goo and merge? Its been long waiting.

@rgommers

rgommers commented:
We may be a week or so away from being able to merge your `spin` PR, so wouldn't it be better to get that done and then tweak this post a bit to say that? It'll read nicer than "almost there, but you can't actually use it yet".

czgdp1807 commented:
Sure. Let's go then. That would be an ideal scenario. <3.

czgdp1807 commented:
@rgommers @pavithraes Since, https://github.com/scipy/scipy/pull/21674 is merged, this blog post is also ready to go. I made last few changes to the content (see, https://github.com/Quansight/Quansight-website/pull/893/commits/5aea04aea9f6e93d87fe1e9459f3d816f586f05d, https://github.com/Quansight/Quansight-website/pull/893/commits/a0cbea575cae32d042990d8eb0f2a4aa5f374270 and https://github.com/Quansight/Quansight-website/pull/893/commits/243286a36bf33131e5f09437c43ab827b45b5bde). 

I will update the date once this PR is approved for merging.

rgommers commented on apps/labs/posts/intel-oneapi-msvc-spin-scipy.md:
```diff
@@ -0,0 +1,187 @@
+In this step, we replaced `gcc` with `icx` and `g++` with `icpx`, and the build successfully passed on Linux.
+
+However, on Windows, we were unable to proceed further due to an unresolved import error with `arpack`.
+Despite several attempts to fix the issue, all solutions led to dead ends. As it was turning into an unbounded amount of work, we decided to pause the effort on Windows and focus on more scoped tasks.
```
```suggestion
Despite several attempts to fix the issue, all solutions led to dead ends. As it was turning into an unbounded amount of work and the SciPy team is planning to rewrite ARPACK completely in a language other than FORTRAN, we decided to pause the effort on Windows and focus on more scoped tasks.
```

rgommers reviewed (approved):
Yes, looks ready now. Thanks @czgdp1807, and thanks for all the help @pavithraes!

@czgdp1807 I made one more suggestion, can you please check it? And then update the publication date?

czgdp1807 commented:
@rgommers @pavithraes I have applied [this suggestion](https://github.com/Quansight/Quansight-website/pull/893#discussion_r2068742756) and updated the publication date. So let's go and merge this.
